File: src/main/java/com/kasisoft/libs/common/functional/KFunction.java
package com.kasisoft.libs.common.functional;

import com.kasisoft.libs.common.*;

import jakarta.validation.constraints.*;

import java.util.function.*;

/**
 * @author daniel.kasmeroglu@kasisoft.com
 */
@FunctionalInterface
public interface KFunction<T, R> {

    R apply(T input) throws Exception;

    @NotNull
    default <V> KFunction<V, R> compose(@NotNull KFunction<? super V, ? extends T> before) {
        return $v -> apply(before.apply($v));
    }

    @NotNull
    default <V> KFunction<T, V> andThen(@NotNull KFunction<? super R, ? extends V> after) {
        return $v -> after.apply(apply($v));
    }

    @NotNull
    default Function<T, R> protect() {
        return $t -> {
            try {
                return apply($t);
            } catch (Exception ex) {
                throw KclException.wrap(ex);
            }
        };
    }

    @NotNull
    static <T> KFunction<T, T> identity() {
        return $t -> $t;
    }

} /* ENDINTERFACE */

File: src/main/kotlin/com/kasisoft/libs/common/functional/KFunction.kt
package com.kasisoft.libs.common.functional;

import com.kasisoft.libs.common.*;

import java.util.function.Function

/**
 * This implementation allows to throw Exceptions (useful on the Java side)
 *
 * @author daniel.kasmeroglu@kasisoft.com
 */
fun interface KFunction<T: Any?, R: Any?> {

    @Throws(Exception::class)
    fun apply(t: T): R

    fun protect(): Function<T, R> =
        Function<T, R> { t: T ->
            try {
                apply(t)
            } catch (ex: Exception) {
                throw KclException.wrap(ex)
            }
        }

} /* ENDINTERFACE */
